certora/specs/MerkleTrees.spec

```diff
@@ -0,0 +1,36 @@
+methods {
+    function newInternalNode(address, address, address, address) external envfree;
+
+    function getRoot(address) external returns address envfree;
+    function getValue(address, address) external returns uint256 envfree;
+    function isEmpty(address, address) external returns bool envfree;
+    function isWellFormed(address, address) external returns bool envfree;
+}
+
+invariant zeroIsEmpty(address tree)
+    isEmpty(tree, 0);
+
+invariant nonEmptyHasValue(address tree, address addr)
+    ! isEmpty(tree, addr) => getValue(tree, addr) != 0
+{ preserved newInternalNode(address _, address parent, address left, address right) {
+    requireInvariant nonEmptyHasValue(tree, left);
+    requireInvariant nonEmptyHasValue(tree, right);
+  }
+}
+
+invariant rootIsZeroOrNotEmpty(address tree)
+    getRoot(tree) == 0 || !isEmpty(tree, getRoot(tree));
+
+invariant wellFormed(address tree, address addr)
+    isWellFormed(tree, addr)
+{ preserved {
+    requireInvariant zeroIsEmpty(tree);
+    requireInvariant wellFormed(tree, addr);
```

Comment:
Don't we require the invariant that we are trying to prove ?

Reply:
You are right, this is not necessary as it is already required by default by the tool

Reply:
ah ok I misunderstood

We are in the `preserved` block, so we are just requiring things to hold before applying the function, so to hold all the time (in which case we wouldn't be proving anything).